python/lsst/pipe/drivers/background.py

```diff
@@ -785,3 +794,52 @@ def run(self, exp):
             detected = mask.array & mask.getPlaneBitMask(['DETECTED']) > 0
             exp.maskedImage.image.array[detected] = smooth.getImage().getArray()[detected]
 
+
+class NanSafeSmoothing:
```

Comment:
>  If you're worried about NaNs, you can always (temporarily?) replace them with zeros when you do the convolution.

This is why I didn't use the existing convolution code. Replacing NaNs with 0 is not equivalent to my code. Some super pixels next to masked pixels are bit far from zero because NaN pixels were originally too high (this is why they are masked).

![Screen Shot 2019-04-02 at 17 17 16](https://user-images.githubusercontent.com/3928908/55386794-388ae880-556b-11e9-8f0c-56c439c77347.png)

A direct convolution in Python is crazy slow, but the targets of this code are relatively small images such as FocalPlaneBackground._values whose dimensions are about 140x150. Convolutions on such images end in about 0.5 second on my mac.

> It looks like you're using class here just as a namespace, which feels dirty. I don't think it adds anything beyond making a stand-alone function with a couple of embedded functions.

I totally agree with you.

Reply:
I figured out the difference between your convolution and the vanilla convolution. This will allow us to condense the code, and probably run it faster:

```
def safeConvolve(array, sigma=2):
    bad = np.isnan(array)
    safe = np.where(bad, 0.0, array)
    convolved = gaussian_filter(safe, sigma, mode="constant", cval=0.0)
    corr = np.where(bad, 0.0, 1.0)
    ones = np.ones_like(array)
    numerator = gaussian_filter(ones, sigma, mode="constant", cval=0.0)
    denominator = gaussian_filter(corr, sigma, mode="constant", cval=0.0)
    return convolved*numerator/denominator
```

This reproduces your result:
![image](https://user-images.githubusercontent.com/1334980/55458164-89fcab80-562f-11e9-835d-9107f6177062.png)

Reply:
Excellent solution.